domain/controllerupgrader/service/agentfinder.go

```diff
@@ -343,5 +406,65 @@ func (a *StreamAgentBinaryFinder) GetHighestPatchVersionAvailableForStream(
 	ctx context.Context,
 	stream agentbinary.Stream,
 ) (semversion.Number, error) {
-	return a.getHighestPatchVersionAvailableForStream(ctx, &stream)
+	if !stream.IsValid() {
+		return semversion.Number{}, errors.Errorf(
+			"agent binary stream %q is not valid", stream,
+		).Add(coreerrors.NotValid)
+	}
+
+	ctrlVersion, err := a.ctrlSt.GetControllerTargetVersion(ctx)
+	if err != nil {
+		return semversion.Zero, errors.Errorf(
+			"getting current controller target agent version: %w", err,
+		)
+	}
+
+	storeBinaries, err := a.ctrlSt.GetAllAgentStoreBinariesForStream(ctx, stream)
```

Comment:
I would cast stream to int here in the service layer, and supply it as int instead of casting in the state layer.